src/celldega/sp/__init__.py

```diff
@@ -0,0 +1,247 @@
+    gdf_points_with_dist[f"distance_to_polygon_{roi_name}"] = gdf_points_with_dist.distance(polygon_geom)
+    print (f"Distances from points to polygon {roi_name} were calculated.")
+    return gdf_points_with_dist, gdf_polygon
+
+def plot_distance_to_roi(
+    gdf_points: gpd.GeoDataFrame,
+    gdf_polygon: gpd.GeoDataFrame,
+    roi_name: str,
+    cmap=cm.plasma,
+    markersize: float = 0.5
+):
+    """
+    Plot the points colored by their distance to the region of interest.
+    
+    Args:
+        gdf_points: GeoDataFrame containing points with distance column.
+        gdf_polygon: GeoDataFrame containing the target polygon.
+        roi_name: Name of the region of interest.
+        cmap: Colormap for distance visualization.
+        markersize: Size of the points in the plot.
+    """
+
+    print (f"Plotting distances from points to polygon {roi_name}...")
+
+    # Normalize distances for color mapping
+    norm = mcolors.Normalize(
+        vmin=gdf_points[f"distance_to_polygon_{roi_name}"].min(),
+        vmax=gdf_points[f"distance_to_polygon_{roi_name}"].max(),
+    )
+    
+    # Map distances to colors
+    gdf_points["color"] = gdf_points[f"distance_to_polygon_{roi_name}"].apply(
+        lambda x: cmap(norm(x))
+    )
+    
+    # Identify points inside the roi
+    polygon = gdf_polygon.geometry.iloc[0]  # Assuming one roi
+    gdf_points["inside_polygon"] = gdf_points["geometry"].apply(
+        lambda x: polygon.contains(x))
+    
+    # Create the plot
+    fig, ax = plt.subplots(figsize=(20, 8))
+    
+    # Plot points outside the roi with gradient color
+    gdf_points[~gdf_points["inside_polygon"]].plot(
+        ax=ax,
+        color=gdf_points[~gdf_points["inside_polygon"]]["color"],
+        aspect=1,
+        legend=True,
+        markersize=markersize,
+    )
+    
+    # Plot the roi as a filled area with label
+    gdf_polygon.plot(
+        ax=ax, color="gray", aspect=1, alpha=0.5, edgecolor="black", linewidth=2
+    )
+    
+    # Annotate the polygon with the ROI name at its centroid
+    centroid = gdf_polygon.geometry.iloc[0].centroid
+    ax.annotate(
+        text=roi_name,
+        xy=(centroid.x, centroid.y),
+        ha="center",
+        va="center",
+        fontsize=12,
+        color="black",
+    )
+    
+    # Add colorbar for the gradient
+    sm = cm.ScalarMappable(cmap=cmap, norm=norm)
+    sm.set_array([])
+    cbar = plt.colorbar(sm, ax=ax, orientation="vertical", pad=0.01, aspect=40)
+    cbar.set_label(f"Distance to ROI {roi_name} (micron)")
+    
+    # Adjust the plot
+    ax.invert_yaxis()
+    plt.title(
+        f"Region Selected Cells Colored Based on Distance to the ROI {roi_name} (micron)"
+    )
+    plt.tight_layout()
+    plt.xlabel("x (micron)")
+    plt.ylabel("y (micron)")
+    plt.show()
+
+
+def calc_gene_expression_by_band(
+    gdf_points: gpd.GeoDataFrame,
+    gdf_bands: gpd.GeoDataFrame,
+    roi_name: str,
+    band_width: float,
+    gene_list: list
```

Comment:
As far as I understand, `gdf_points` contains a subset of genes from the single cell data. Can we modify this so that we pass in as an argument the entire AnnData and/or a cbg matrix so that we calculate the average expression of all genes along these gradient bands?

Reply:
fixed.